### PR TITLE
Update java client

### LINKED ISSUE #15120: Support `UpdateJobTimeout` RPC in Java Client
## Description

Add `Job Update Timeout` command to the Java Client.
This command is very similar to the `Job Update Retries` command.

### PR DESCRIPTION
**Pull request by ana-vinogradova-camunda**
## Description
Added Job Update Timeout command to the Java Client:
- added `newUpdateTimeoutCommand ` to Zeebe client and Zeebe client implementation
- added `JobUpdateTimeoutCommandImpl `
- added the response `UpdateTimeoutJobResponse ` and implementation
- added the test `JobUpdateTimeoutTest `

## Related issues

closes #15120

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: clients/java/src/main/java/io/camunda/zeebe/client/ZeebeClient.java
Comment (tmetzke) [inline clients/java/src/main/java/io/camunda/zeebe/client/ZeebeClient.java:393]: 💭 It might be interesting for users to note any specifics here, e.g. does this add to the existing timeout or replace it, if it replaces it, does it start from now or from when the related command is processed?

File: clients/java/src/main/java/io/camunda/zeebe/client/api/command/UpdateTimeoutJobCommandStep1.java
Comment (tmetzke) [inline clients/java/src/main/java/io/camunda/zeebe/client/api/command/UpdateTimeoutJobCommandStep1.java:28]: 💭 It might be interesting for users to note any specifics here, e.g. does this add to the existing timeout or replace it, if it replaces it, does it start from now or from when the related command is processed?

**Review by tmetzke — CHANGES_REQUESTED** [review]
Really nice work, @ana-vinogradova-camunda 🙌 

👍 The code looks really good and clean, I have nothing to add there.
🔧 I've added a thought for the API docs to improve the user experience. Let me know what you think.

Really well done! 👏

File: clients/java/src/main/java/io/camunda/zeebe/client/ZeebeClient.java
Comment (ana-vinogradova-camunda) [reply clients/java/src/main/java/io/camunda/zeebe/client/ZeebeClient.java:393]: I think this is a very useful addition to the documentation, will update it now.

File: clients/java/src/main/java/io/camunda/zeebe/client/api/command/UpdateTimeoutJobCommandStep1.java
Comment (ana-vinogradova-camunda) [reply clients/java/src/main/java/io/camunda/zeebe/client/api/command/UpdateTimeoutJobCommandStep1.java]: Changes are ready for review.

File: clients/java/src/main/java/io/camunda/zeebe/client/ZeebeClient.java
Comment (ana-vinogradova-camunda) [reply clients/java/src/main/java/io/camunda/zeebe/client/ZeebeClient.java]: Changes are ready for review.

File: clients/java/src/main/java/io/camunda/zeebe/client/impl/response/UpdateTimeoutJobResponseImpl.java
Comment (nicpuppa) [inline clients/java/src/main/java/io/camunda/zeebe/client/impl/response/UpdateTimeoutJobResponseImpl.java:23]: ❓ We should provide a response to the client ?

File: clients/java/src/test/java/io/camunda/zeebe/client/job/JobUpdateTimeoutTest.java
Comment (nicpuppa) [inline clients/java/src/test/java/io/camunda/zeebe/client/job/JobUpdateTimeoutTest.java:29]: 🔧 We should add also some integration test, you can find some example [here](https://github.com/camunda/zeebe/tree/main/qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command)

**Review by nicpuppa — COMMENTED** [review]
Hey @ana-vinogradova-camunda good job 💪  I really like the javadoc. I left some comments, please have a look 👀 

FYI you can use `--amend` [option](https://www.atlassian.com/git/tutorials/rewriting-history) when committing the changes, so instead of 4 commit (like in this case) with the same message you made only one commit at the end. Remember to force push if you use this option 😆

File: clients/java/src/main/java/io/camunda/zeebe/client/impl/response/UpdateTimeoutJobResponseImpl.java
Comment (tmetzke) [reply clients/java/src/main/java/io/camunda/zeebe/client/impl/response/UpdateTimeoutJobResponseImpl.java:23]: If I'm not mistaken, the [response is always empty](https://github.com/camunda/zeebe/blob/main/gateway-protocol/src/main/proto/gateway.proto#L556-L557)? Would there be anything we need to add here in that case?

Comment (nicpuppa) [reply clients/java/src/main/java/io/camunda/zeebe/client/impl/response/UpdateTimeoutJobResponseImpl.java:23]: Yes, I saw that. My question is more general 😄 We should provide some infos in the response to the clients ?

Comment (korthout) [reply clients/java/src/main/java/io/camunda/zeebe/client/impl/response/UpdateTimeoutJobResponseImpl.java:23]: 💭 It might be useful for the user to know the new deadline, but if we want to add this it can also be done in a later iteration.

Comment (tmetzke) [reply clients/java/src/main/java/io/camunda/zeebe/client/impl/response/UpdateTimeoutJobResponseImpl.java:23]: I tend to agree with @korthout here. It's something we could add later if we see a need or get such feedback.

File: clients/java/src/main/java/io/camunda/zeebe/client/api/command/UpdateTimeoutJobCommandStep1.java
Comment (tmetzke) [reply clients/java/src/main/java/io/camunda/zeebe/client/api/command/UpdateTimeoutJobCommandStep1.java:32]: I like that idea a lot 👍 Would you support both options here by providing both methods? Is this something we do in other commands as well?

File: clients/java/src/main/java/io/camunda/zeebe/client/impl/response/UpdateTimeoutJobResponseImpl.java
Comment (ana-vinogradova-camunda) [reply clients/java/src/main/java/io/camunda/zeebe/client/impl/response/UpdateTimeoutJobResponseImpl.java:23]: As per Remco's advise the response is currently empty (the same as with update retries) but if it needs to be changed I can do so.

File: clients/java/src/main/java/io/camunda/zeebe/client/api/command/UpdateTimeoutJobCommandStep1.java
Comment (korthout) [reply clients/java/src/main/java/io/camunda/zeebe/client/api/command/UpdateTimeoutJobCommandStep1.java:32]: Yup, I'd overload the method

https://github.com/camunda/zeebe/blob/fde5ae4d20eb8e9372195fa25f5acd9dc8404ed8/clients/java/src/main/java/io/camunda/zeebe/client/api/worker/JobWorkerBuilderStep1.java#L67-L93

File: clients/java/src/test/java/io/camunda/zeebe/client/job/JobUpdateTimeoutTest.java
Comment (ana-vinogradova-camunda) [reply clients/java/src/test/java/io/camunda/zeebe/client/job/JobUpdateTimeoutTest.java:29]: Done and pushed, thank you

**Review by nicpuppa — APPROVED** [review]
Thanks @ana-vinogradova-camunda for applied my suggestion. 

LGTM 🚀 

Before merging we need to wait @tmetzke approval 😄

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/UpdateJobTimeoutTest.java
Comment (tmetzke) [inline qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/UpdateJobTimeoutTest.java]: 💭 I am wondering if this is too error-prone and will lead to flaky results. We're relying on timing here, if I understand this right. Could we also fetch the "old" deadline, increase it with the update method, and assert that the new one is greater?

I might also simply misunderstand the code here. In that case, just move on and ignore this.

**Review by tmetzke — APPROVED** [review]
👍 Great, @ana-vinogradova-camunda, thanks for looking into all those suggestions!

💭 I added some thoughts regarding the integration tests. Feel free to ignore them if my assumption is incorrect.

Approving already so you can make progress in case the test is fine as it is.

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/UpdateJobTimeoutTest.java
Comment (nicpuppa) [reply qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/UpdateJobTimeoutTest.java]: Good point @tmetzke

Comment (ana-vinogradova-camunda) [reply qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/UpdateJobTimeoutTest.java]: @tmetzke I like that idea, let me make a change.

Comment (ana-vinogradova-camunda) [reply qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/UpdateJobTimeoutTest.java]: @tmetzke I could set the initial Job deadline to be (example) 14 minutes away from now and then update the timeout to 15 minutes away from now. That would take current time and add 15 minutes to it. I could compare that the new deadline is greater than the old one then. I could take similar approach when testing reducing the job timeout as well.

Comment (ana-vinogradova-camunda) [reply qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/UpdateJobTimeoutTest.java]: Change is ready for review.

**Review by tmetzke — APPROVED** [review]
👍 Great, @ana-vinogradova-camunda, that should remove potential flakiness.
💭 I added two more thoughts below but they are more on the "style" side of things. You can also ignore them if you feel like it's becoming too much 🙂

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/UpdateJobTimeoutTest.java
Comment (tmetzke) [reply qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/UpdateJobTimeoutTest.java:147]: Forgot to mention the motivation: I think (but I am not entirely sure, to be honest) we iterate through the records twice to first check if the job exists (for the assertion) and then for fetching is and returning it's deadline.

Comment (ana-vinogradova-camunda) [reply qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/UpdateJobTimeoutTest.java:147]: I think that both things are needed - the first assertions verifies that JobIntent.UPDATE_TIMEOUT exits for the specified job key, the second part (line 143 to 147) looks for JobIntent.TIMEOUT_UPDATED record and extracts the `deadline` value.

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/UpdateJobTimeoutTest.java
Comment (ana-vinogradova-camunda) [reply qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/UpdateJobTimeoutTest.java]: Done!

Comment (tmetzke) [reply qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/UpdateJobTimeoutTest.java]: 👍

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/UpdateJobTimeoutTest.java
Comment (tmetzke) [reply qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/UpdateJobTimeoutTest.java:147]: Oh damn, I totally missed the different intents 😄🤦 Thanks for pointing that out 👍 Good to go from my end 🙂

**Review by tmetzke — APPROVED** [review]
👍 Awesome, good to go from my side :shipit:

**Comment by ana-vinogradova-camunda** [comment]
bors merge

**Pull request merged**
